widgets/heatmap/heatmap.go

```diff
@@ -89,42 +144,161 @@ func (hp *HeatMap) ClearYLabels() {
 }
 
 // axesDetails determines the details about the X and Y axes.
 func (hp *HeatMap) axesDetails(cvs *canvas.Canvas) (*axes.XDetails, *axes.YDetails, error) {
-	return nil, nil, errors.New("not implemented")
+	hp.cellWidthAdaptive(cvs)
+
+	yd, err := axes.NewYDetails(hp.yLabels)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	xd, err := axes.NewXDetails(yd.End, hp.xLabels, hp.opts.cellWidth)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return xd, yd, nil
 }
 
 // Draw draws cells, X labels and Y labels as HeatMap.
 // Implements widgetapi.Widget.Draw.
 func (hp *HeatMap) Draw(cvs *canvas.Canvas, meta *widgetapi.Meta) error {
-	return errors.New("not implemented")
+	hp.mu.Lock()
+	defer hp.mu.Unlock()
+
+	hp.lastWidth = cvs.Area().Dx()
+	hp.lastHeight = cvs.Area().Dy()
+	// Check if the canvas has enough area to draw HeatMap.
+	needAr, err := area.FromSize(hp.minSize())
+	if err != nil {
+		return err
+	}
+	if !needAr.In(cvs.Area()) {
+		return draw.ResizeNeeded(cvs)
+	}
+
+	xd, yd, err := hp.axesDetails(cvs)
+	if err != nil {
+		return err
+	}
+	err = hp.drawCells(cvs, yd)
+	if err != nil {
+		return err
+	}
+
+	return hp.drawLabels(cvs, xd, yd)
 }
 
 // drawCells draws m*n cells (rectangles) representing the stored values.
 // The height of each cell is 1 and the default width is 3.
-func (hp *HeatMap) drawCells(cvs *canvas.Canvas, xd *axes.XDetails, yd *axes.YDetails) error {
-	return errors.New("not implemented")
+func (hp *HeatMap) drawCells(cvs *canvas.Canvas, yd *axes.YDetails) error {
+	for i := 0; i < len(hp.values); i++ {
+		for j := 0; j < len(hp.values[0]); j++ {
+			startX := yd.Start.X + axes.AxisWidth + j*hp.opts.cellWidth
+			startY := yd.Labels[i].Pos.Y
+
+			endX := startX + hp.opts.cellWidth
+			endY := startY + 1
+
+			rect := image.Rect(startX, startY, endX, endY)
+			color := hp.getCellColor(hp.values[i][j])
+
+			if err := cvs.SetAreaCells(rect, hp.opts.cellChar, cell.BgColor(color)); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
 }
 
 // drawAxes draws X labels (under the cells) and Y Labels (on the left side of the cell).
 func (hp *HeatMap) drawLabels(cvs *canvas.Canvas, xd *axes.XDetails, yd *axes.YDetails) error {
-	return errors.New("not implemented")
+	if !hp.opts.hideYLabels {
+		for _, l := range yd.Labels {
+			if err := draw.Text(cvs, l.Text, l.Pos,
+				draw.TextCellOpts(hp.opts.yLabelCellOpts...),
+			); err != nil {
+				return fmt.Errorf("failed to draw the Y labels: %v", err)
+			}
+		}
+	}
+
+	if !hp.opts.hideXLabels {
+		for _, l := range xd.Labels {
+			if err := draw.Text(cvs, l.Text, l.Pos, draw.TextCellOpts(hp.opts.xLabelCellOpts...)); err != nil {
+				return fmt.Errorf("failed to draw the X labels: %v", err)
+			}
+		}
+	}
+	return nil
+}
+
+// minCellWidth is the minimum width of each cell in the heat map.
+const minCellWidth = 3
+
+// cellWidthAdaptive determines the width of a single cell (grid) based the canvas.
+func (hp *HeatMap) cellWidthAdaptive(cvs *canvas.Canvas) {
```

Comment:
I am wondering if we could improve the name `cellWidthAdaptive`. How does `adaptCellWidth` sound?